### PR TITLE
Models team update: anomaly detection pipeline, benchmarks, documentation, and cleanup

### PR DESCRIPTION
**Pull request by lukitasxue**
# Models Team PR Materials

## PR title

Models team update: anomaly detection pipeline, benchmarks, documentation, and cleanup

---

## Commit message

feat(data_science): add models benchmark pipeline, docs, and folder cleanup

---

## Summary

This PR merges the Models team work into the main repository.

It includes the anomaly detection pipeline, multiple detector implementations, benchmark/evaluation functionality, output report generation, NAB support, documentation, and folder cleanup for handover.

The goal is to make the Models team work usable, testable, and understandable for the wider project team, including Backend, Frontend, mentors, and future students.

## High-level pipeline

The Models pipeline now follows this structure:

    dataset
    - preprocessor.py
    - anomaly_injector.py or nab_loader.py
    - detectors/
    - evaluator.py
    - roc_plotter.py / report_output.py
    - outputs/

The system can load IoT-style sensor data, preprocess it, run anomaly detectors, evaluate predictions, and generate CSV, JSON, plot, and text report outputs.

### Anomaly detection pipeline

Added and updated the main pipeline flow in `data_science/pipeline.py`.

The pipeline can:

- load sensor datasets
- preprocess numeric time-series data
- run all registered detectors
- validate detector outputs
- evaluate detector performance
- save benchmark results
- generate plots and reports

### Detector implementations

Added/organised multiple anomaly detectors under `data_science/detectors/`, including:

- `PcaADDetector`
- `OCSVMDetector`
- `LevelShiftADDetector`
- `VolatilityShiftADDetector`
- `QuantileADDetector`
- `ECODDetector`
- `COPODDetector`
- `InterQuartileRangeADDetector`
- `LOFDetector`
- `ThresholdADDetector`

These include ADTK-based, PyOD-based, scikit-learn-based, and custom statistical detectors.

### Benchmark modes

Added support for three benchmark modes:

1. Synthetic benchmark
   - uses `datasets/complex.csv`
   - injects synthetic anomalies
   - evaluates detectors against known injected labels

2. Train/test benchmark
   - splits data into train and test sections
   - fits trainable detectors on training data where supported
   - injects anomalies into the test split
   - evaluates on unseen test data

3. NAB benchmark
   - uses the public Numenta Anomaly Benchmark dataset
   - loads official NAB anomaly windows
   - evaluates detectors against real labelled anomaly periods

### Synthetic anomaly injection

Added/updated `anomaly_injector.py`, which supports:

- point spike anomalies
- level shift anomalies
- volatility shift anomalies
- combined anomaly injection through `inject_all()`

### NAB support

Added `nab_loader.py` to load NAB anomaly windows from `combined_windows.json` and align them with the dataset timestamp index.

Note: the current NAB evaluation uses pointwise precision, recall, F1, and AUC-ROC. It does not yet implement the official NAB scoring profiles.

### Evaluation metrics

Added/updated evaluator support for:

- precision
- recall
- F1
- AUC-ROC
- number of predicted anomalies
- number of actual anomalies
- true positives
- false positives
- true negatives
- false negatives
- false positive rate
- false negative rate

### Report and output generation

Added/updated output generation through:

- `report_output.py`
- `roc_plotter.py`
- `final_report.py`
- `run_all_benchmarks.py`

The pipeline now generates files such as:

- `benchmark_results.csv`
- `benchmark_results.json`
- `metrics_comparison.png`
- `runtime_comparison.png`
- `confusion_summary.png`
- `model_flags_timeline.png`
- `roc_curves.png`
- `benchmark_report_summary.txt`
- `final_benchmark_summary.csv`
- `final_benchmark_report.txt`
- `final_f1_comparison.png`
- `final_auc_comparison.png`

### Documentation and handover

Added a full documentation structure for the Models team work:

- `data_science/README.md`
- `data_science/BENCHMARKING_README.md`
- `data_science/outputs/README.md`
- `data_science/detectors/README.md`
- `data_science/detectors/ADTK_DETECTORS_README.md`

Added detector-specific README files for:

- PCAAD
- OCSVM
- LOF
- ECOD
- COPOD
- ThresholdAD
- InterQuartileRangeAD

These README files explain the system structure, benchmark commands, detector behaviour, known limitations, outputs, and handover guidance.

### Folder cleanup and archive

The `data_science/` folder was cleaned up to separate active pipeline files from older prototype work.

Old prototype files were moved into `data_science/archive/`.

This includes older files related to:

- previous algorithm prototypes
- old development/API scripts
- old visualisation scripts
- old generated/storage files
- older LSTM/experimental work
- old datasets not used by the current benchmark pipeline

No files were deleted. They were moved to archive to preserve historical evidence from previous generations while keeping the active Models system easier to review and maintain.

## How to test

From inside `data_science/`, install dependencies:

    pip install -r requirements.txt

Run synthetic benchmark:

    python pipeline.py datasets/complex.csv --benchmark

Run train/test benchmark:

    python pipeline.py datasets/complex_clean.csv --benchmark --train-test

Run NAB benchmark:

    python pipeline.py "https://raw.githubusercontent.com/numenta/NAB/master/data/realKnownCause/machine_temperature_system_failure.csv" --benchmark --label-source nab --nab-label-file "https://raw.githubusercontent.com/numenta/NAB/master/labels/combined_windows.json" --nab-dataset-key "realKnownCause/machine_temperature_system_failure.csv" --output-dir outputs/nab_benchmark

Run all benchmarks:

    python run_all_benchmarks.py

Run injector tests:

    python -m pytest tests/test_injector.py

## Validation

The following benchmark modes were tested after the folder cleanup/refactor:

- synthetic benchmark
- train/test benchmark
- NAB benchmark

The benchmark runs completed and generated the expected output files under `data_science/outputs/`.

## Notes for reviewers

Some warnings may appear during benchmark runs:

- ADTK/pandas `FutureWarning`s
- matplotlib `set_ticklabels()` warnings
- ROC/AUC warnings in some NAB cases where only one label class is present in a scoring slice

These warnings do not currently stop the benchmark outputs from being generated.

The main purpose of this PR is to merge the Models team pipeline, detectors, benchmark system, documentation, and cleanup into the main repo. It is not intended to finalise production Backend/Frontend integration yet.

## Suggested review focus

Please review:

- whether the `data_science/` folder structure is clear
- whether the README documentation is understandable
- whether the benchmark commands run correctly
- whether detector outputs follow the expected pipeline interface
- whether the archived files are correctly separated from the active system
- whether the output structure is useful for future Backend/Frontend integration and handover

---

## Team message

Hey team, I pushed the Models team work and opened a PR/merge request for it.

This is not just today’s documentation cleanup, it includes the full Models team work we have been building: the anomaly detection pipeline, detectors, benchmark modes, evaluation logic, report outputs, documentation, and folder organisation.

Most of the details are explained in the PR description and the README files, but here is the high-level guide for reviewing it.

The main Models pipeline is inside `data_science/`.

High-level flow:

    dataset
    - preprocessor.py
    - anomaly_injector.py or nab_loader.py
    - detectors/
    - evaluator.py
    - roc_plotter.py / report_output.py
    - outputs/

In simple terms, the pipeline loads sensor data, preprocesses it, runs multiple anomaly detectors, evaluates their predictions, and saves benchmark outputs such as CSV, JSON, plots, and text summaries into the `outputs/` folder.

The main documentation files are:

    data_science/README.md
    data_science/BENCHMARKING_README.md
    data_science/outputs/README.md
    data_science/detectors/README.md
    data_science/detectors/ADTK_DETECTORS_README.md

There are also detector-specific README files for PCAAD, OCSVM, LOF, ECOD, COPOD, ThresholdAD, and InterQuartileRangeAD.

Before running anything, please go into `data_science/` and install the requirements:

    pip install -r requirements.txt

Main commands to test:

    python pipeline.py datasets/complex.csv --benchmark

    python pipeline.py datasets/complex_clean.csv --benchmark --train-test

    python pipeline.py "https://raw.githubusercontent.com/numenta/NAB/master/data/realKnownCause/machine_temperature_system_failure.csv" --benchmark --label-source nab --nab-label-file "https://raw.githubusercontent.com/numenta/NAB/master/labels/combined_windows.json" --nab-dataset-key "realKnownCause/machine_temperature_system_failure.csv" --output-dir outputs/nab_benchmark

Or to run all benchmark modes together:

    python run_all_benchmarks.py

There is also a current test file for the anomaly injector:

    python -m pytest tests/test_injector.py

You may see some ADTK/pandas warnings, matplotlib warnings, or ROC/AUC warnings during benchmark runs. These are expected at the moment and do not necessarily mean the pipeline failed. The important part is that the benchmark completes and saves the outputs into the correct `outputs/` folder.

I also moved a bunch of older prototype files into `data_science/archive/`. These were not being used by the current active pipeline anymore. I checked references before moving them, so the current benchmark system should not break. I did not delete them, they are preserved as historical evidence from previous project generations.

For review, the main things to check are:

1. Does the folder structure make sense?
2. Are the README files clear enough for handover?
3. Do the benchmark commands run correctly?
4. Are the detector files and benchmark outputs understandable?
5. Is the archive separation okay for old unused files?

All deeper details are in the PR description and README files.

### CODE REVIEW

**Comment by senuradp** [comment]
Had a proper look through the PR description, project structure, and the overall models pipeline organisation.
 
This is a really solid milestone for the Models team. The work has clearly moved beyond prototype experimentation into a much more structured and reusable benchmarking pipeline.
 
The overall separation of concerns looks well thought out. Splitting preprocessing, anomaly injection/NAB loading, detector implementations, evaluation logic, plotting, reporting outputs, and documentation into clear modules makes the workflow much easier to understand, maintain, and hand over. The archive separation for older prototype work also makes sense since it keeps historical work preserved without cluttering the active pipeline.
 
I also like the documentation effort here. Having dedicated README files for benchmarking, detectors, outputs, and detector-specific implementations makes onboarding and future continuation much smoother, especially for a capstone project where maintainability and handover matter.
 
From a Data Science integration perspective, the benchmark pipeline structure looks consistent and reusable, and the support for multiple benchmark modes/datasets gives good flexibility for future testing and model comparison.
 
No blocking concerns from my side at this stage. Great work getting the Models pipeline to main PR stage
